File: aoc2020/day6/part2.ts
async function main() {
  const file = Bun.file("./input.txt");
  const arrBuffer = await file.arrayBuffer();
  const buffer = Buffer.from(arrBuffer);
  const data = buffer.toString();

  const splitGroups = (input: string): string[][] => {
    return input
      .trim()
      .split("\n\n")
      .map((grp) => grp.split("\n"));
  };

  const countCommonYesAns = (group: string[]): number => {
    const ansSets = group.map((ans) => new Set(ans));
    const commonAns = ansSets.reduce((common, set) => {
      return new Set([...common].filter((ans) => set.has(ans)));
    }, ansSets[0]);

    return commonAns.size;
  };

  const countTotalCommonYesAnswers = (input: string): number => {
    const groups = splitGroups(input);
    return groups.reduce(
      (total, group) => total + countCommonYesAns(group),
      0
    );
  };

  console.log(countTotalCommonYesAnswers(data));
}

await main();

File: aoc2020/day6/part2.py
def split_groups(input_str: str) -> list[list[str]]:
    return [group.split("\n") for group in input_str.strip().split("\n\n")]

def count_common_yes_ans(group: list[str]) -> int:
    # Convert each answer to a set of characters
    ans_sets = [set(ans) for ans in group]

    # Find intersection of all sets in the group
    # Using * to unpack the list of sets for set.intersection()
    common_ans = set.intersection(*ans_sets)
    return len(common_ans)

def count_total_common_yes_answers(input_str: str) -> int:
    groups = split_groups(input_str)
    return sum(count_common_yes_ans(group) for group in groups)

def main():
    with open("input.txt", "r") as file:
        data = file.read()
    print(count_total_common_yes_answers(data))

if __name__ == "__main__":
    main()
